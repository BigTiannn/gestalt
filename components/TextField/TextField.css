.TextField {
    background-color: #f3f3f3;
    background-image: linear-gradient(#f0f0f0, #f5f5f5);
    box-sizing: border-box;
    width: 100%;
    border-radius: 3px;
    border: 1px solid #d6d3ce;
    box-shadow: 0 0 3px rgba(0, 0, 0, 0.12) inset;
    color: #333;
    font-size: 14px;
    padding: 7px;
}

.TextField--isErrored {
    background: #e6abad;
    border-color: #bd081c;
}

.TextField--isFocused,
.TextField:focus {
    background: #fff;
    box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.15) inset;
    outline: #d1d1d1 auto 1px;
}

.TextField--isDisabled,
.TextField:disabled {
    background-color: #f8f8f8;
    background-image: linear-gradient(#f5f5f5, #fafafa);
    border-color: #dcdcdc;
    box-shadow: 0 0 5px rgba(0, 0, 0, 0.12) inset;
    color: #999;
    cursor: default;
}
