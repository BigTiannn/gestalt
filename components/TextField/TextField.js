import React, { Component, PropTypes } from 'react';
import classNames from 'classnames';
import styles from './TextField.css';

export default class TextField extends Component {

    static propTypes = {
        classNames: PropTypes.arrayOf(PropTypes.string),
        disabled: PropTypes.bool,
        focused: PropTypes.bool,
        id: PropTypes.string,
        name: PropTypes.string,
        onBlur: PropTypes.func,
        onChange: PropTypes.func,
        onClick: PropTypes.func,
        onValidation: PropTypes.func,
        placeholder: PropTypes.string,
        type: PropTypes.oneOf(['url', 'text', 'password']),
        url: PropTypes.bool,
        validators: PropTypes.arrayOf(PropTypes.func),
        value: PropTypes.string,
    }

    static defaultProps = {
        validators: [],
        type: 'text'
    }

    constructor(props, context) {
        super(props, context);

        this.state = {
            isFocused: this.props.focused || false,
            value: this.props.value,
            hasChanged: this.props.value !== undefined
        };
    }

    componentDidMount() {
        if (this.state.isFocused) {
            this._input.focus();
        }
    }

    componentWillReceiveProps(nextProps) {
        this.setState({value: nextProps.value});
    }

    handleChange(e) {
        if (this.props.onChange) {
            this.props.onChange(e.target.value);
        }
        let validatorErrors = this.runValidators(e.target.value);
        if (this.props.onValidation) {
            this.props.onValidation(validatorErrors);
        }
        this.setState({value: e.target.value, hasChanged: true});
    }

    getId() {
        return this.props.id;
    }

    getValue() {
        return this.state.value;
    }

    runValidators(value) {
        let validatorErrors = [];
        if (this.state.hasChanged) {
            for (let validator of this.props.validators){
                let data = validator(value);
                validatorErrors = validatorErrors.concat(data['errors']);
            }
        }
        return validatorErrors;
    }

    render() {
        const validatorErrors = this.runValidators(this.state.value);
        const modifiers = {};
        modifiers[styles['TextField--has-error']] = this.state.hasChanged && validatorErrors.length > 0;

        const myClasses = classNames(
            styles.TextField,
            this.props.classNames,
            modifiers
        );

        return (
            <input
                className={myClasses}
                disabled={this.props.disabled ? 'disabled' : ''}
                id={this.props.id}
                name={this.props.name}
                onBlur={this.props.onBlur ? this.props.onBlur.bind(this) : null}
                onChange={this.handleChange.bind(this)}
                onClick={this.props.onClick ? this.props.onClick.bind(this) : null}
                placeholder={this.props.placeholder}
                ref={(comp) => this._input = comp}
                type={this.props.type}
                value={this.state.value}
            />
        );
    }
}
