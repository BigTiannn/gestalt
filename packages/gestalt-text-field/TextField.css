.common {
  composes: bg-light-gray from "gestalt-colors/Colors.css";
  composes: bold from "gestalt-typography/Typography.css";
  composes: border-box from "gestalt-borders/Borders.css";
  composes: col-12 from "gestalt-column/Column.css";
  composes: dark-gray from "gestalt-colors/Colors.css";
  composes: px1 from "gestalt-whitespace/Whitespace.css";
  composes: py2 from "gestalt-whitespace/Whitespace.css";
  composes: rounded from "gestalt-borders/Borders.css";
  composes: text-m from "gestalt-typography/Typography.css";
  cursor: text;
  outline: 0;
}

.textfield {
  composes: common;
  composes: border from "gestalt-borders/Borders.css";
}

.textfield:focus {
  background-color: #fff;
}

.errored {
  composes: common;
  border: 1px solid #bd081c;
}

.errored:focus {
  background-color: #fff;
}
