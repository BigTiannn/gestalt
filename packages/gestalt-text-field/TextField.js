// @flow

import React, { PropTypes } from 'react';
import classnames from 'classnames/bind';
import styles from './TextField.css';

const cx = classnames.bind(styles);

function handleChange(e: Event, onChange) {
  if (e.target instanceof HTMLInputElement) {
    onChange(e.target.value);
  }
}

type Props = {
  hasError?: boolean,
  id: string,
  onChange: (value: string) => void,
  placeholder?: string,
  type?: 'email' | 'password' | 'text' | 'url',
  value?: string,
};

export default function TextField(props: Props) {
  const {
    hasError = false,
    id,
    onChange,
    placeholder,
    type = 'text',
    value,
  } = props;

  const classes = cx({
    textfield: !hasError,
    errored: hasError,
  });

  return (
    <div>
      <input
        className={classes}
        id={id}
        onChange={(e: Event) => { handleChange(e, onChange); }}
        placeholder={placeholder}
        type={type}
        value={value}
      />
    </div>
  );
}

TextField.propTypes = {
  hasError: PropTypes.bool,
  id: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.oneOf(['email', 'password', 'text', 'url']),
  value: PropTypes.string,
};
