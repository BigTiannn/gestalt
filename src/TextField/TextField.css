.common {
  composes: bg-light-gray dark-gray from "../Colors.css";
  composes: bold text-md from "../Typography.css";
  composes: border-box from "../Layout.css";
  composes: col-12 from "../Column/Column.css";
  composes: px1 py2 from "../Whitespace.css";
  composes: rounded from "../Borders.css";
  cursor: text;
  outline: 0;
}

.textfield {
  composes: common;
  composes: border from "../Borders.css";
}

.textfield:focus {
  background-color: #fff;
}

.errored {
  composes: common;
  border: 1px solid #fab904;
}

.errored:focus {
  background-color: #fff;
}
