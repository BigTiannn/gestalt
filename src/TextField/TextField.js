// @flow

import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import ErrorFlyout from '../ErrorFlyout/ErrorFlyout';
import styles from './TextField.css';

type State = {
  focused: boolean,
  errorIsOpen: boolean,
};

export default class TextField extends Component {
  static propTypes = {
    errorMessage: PropTypes.string,
    hasError: PropTypes.bool,
    id: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    placeholder: PropTypes.string,
    type: PropTypes.oneOf(['email', 'password', 'text', 'url']),
    value: PropTypes.string,
  };

  static defaultProps = {
    hasError: false,
    type: 'text',
  };

  state:State = {
    focused: false,
    errorIsOpen: false,
  };

  props: {|
    errorMessage?: string,
    hasError?: bool,
    id: string,
    onBlur?: (value: string) => void,
    onChange: (value: string) => void,
    onFocus?: (value: string) => void,
    placeholder?: string,
    type?: 'email' | 'password' | 'text' | 'url',
    value?: string,
  |};

  handleChange = (e: Event) => {
    if (e.target instanceof HTMLInputElement) {
      this.props.onChange(e.target.value);
    }
  }

  handleBlur = (e: Event) => {
    this.setState({ errorIsOpen: false });
    if (e.target instanceof HTMLInputElement && this.props.onBlur) {
      this.props.onBlur(e.target.value);
    }
  }

  handleFocus = (e: Event) => {
    this.setState({ errorIsOpen: true });
    if (e.target instanceof HTMLInputElement && this.props.onFocus) {
      this.props.onFocus(e.target.value);
    }
  }

  render() {
    const {
      errorMessage,
      hasError,
      id,
      placeholder,
      type,
      value,
    } = this.props;

    const classes = classnames(styles.textField, {
      [styles.normal]: !errorMessage,
      [styles.errored]: hasError || errorMessage,
    });

    const textField = (onBlur, onFocus) => (
      <input
        aria-describedby={errorMessage && this.state.focused ? `${id}-gestalt-error` : null}
        aria-invalid={(errorMessage || hasError) ? 'true' : 'false'}
        className={classes}
        id={id}
        onBlur={onBlur}
        onChange={this.handleChange}
        onFocus={onFocus}
        placeholder={placeholder}
        type={type}
        value={value}
      />
    );

    return errorMessage ? (
      <ErrorFlyout
        id={`${id}-gestalt-error`}
        idealDirection="right"
        isOpen={this.state.errorIsOpen}
        message={errorMessage}
        onDismiss={() => this.setState({ errorIsOpen: false })}
        size="sm"
        trigger={textField(this.handleBlur, this.handleFocus)}
      />
    ) : textField();
  }
}
