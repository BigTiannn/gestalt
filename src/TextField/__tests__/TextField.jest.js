/* eslint-env jest */
import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import TextField from '../TextField';

describe('TextField', () => {
  it('Renders an ErrorFlyout if one is passed in', () => {
    const wrapper = shallow(
      <TextField
        errorMessage="test"
        id="test"
        onChange={jest.fn()}
      />
    );
    expect(wrapper.find('ErrorFlyout').length).toEqual(1);
  });

  it('Does not render an ErrorFlyout when errorMessage is null', () => {
    const wrapper = shallow(
      <TextField
        id="test"
        onChange={jest.fn()}
      />
    );
    expect(wrapper.find('ErrorFlyout').length).toEqual(0);
  });

  it('TextField normal', () => {
    const tree = create(
      <TextField
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('TextField with error', () => {
    const tree = shallow(
      <TextField
        errorMessage="error message"
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });
});
