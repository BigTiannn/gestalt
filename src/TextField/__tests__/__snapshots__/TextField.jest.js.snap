exports[`TextField TextField normal 1`] = `
<input
  aria-describedby={null}
  aria-invalid="false"
  className="textField normal"
  id="test"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder={undefined}
  type="text"
  value={undefined} />
`;

exports[`TextField TextField with error 1`] = `"<div><div><input type=\"text\" aria-invalid=\"true\" class=\"textField errored\" id=\"test\"/></div><div></div></div>"`;
